src/main/java/com/google/devtools/build/lib/cmdline/RepositoryMapping.java

```diff
@@ -134,6 +157,34 @@ public RepositoryMapping composeWith(RepositoryMapping other) {
       RepositoryName mappedName = other.get(entry.getValue().getName());
       entries.put(entry.getKey(), mappedName.isVisible() ? mappedName : entry.getValue());
     }
-    return createInternal(entries, null);
+    return new RepositoryMapping(entries, null);
+  }
+
+  /**
+   * Returns a new {@link RepositoryMapping} instance with identical contents, except that the
+   * inverse mapping is cached, causing {@link #getInverse} to be much more efficient. This is
+   * particularly important for the main repo mapping, as it's often used to generate display-form
```

Comment:
This comment and the name make it sound as if this could be applied to any other repo, but the class is called `MainRepoMapping`. Maybe call it `CachedInverseRepoMapping` instead?

Reply:
fixed to just use an inline subclass instead